Milestone Notification text styles and confetti decoration

frosty opened the pull request:
This PR makes some changes to the way 'badge' notifications are styled, as part of the Milestone Notifications project. I've written the changes to include all badge notifications, not just view milestones. The key changes are:

- Milestone notifications now show a confetti image behind the main number image. This is added via a new `backgroundImage` property on the image cell.
- The first text block in a notification now has a new 'title' text style, of the system serif font in the `title1` size.
- Subsequent text blocks have a slight tweak, using `body` style in the standard system font.
- The code added to WPStyleGuide+Notifications and BadgeContentStyles will become a bit cleaner once we remove the feature flag.
- I also took the opportunity to group some of the assets in the main asset catalogue in a Notifications folder.

| **Before**  |  **After** |
|:-:|:-:|
| ![IMG_0031](https://user-images.githubusercontent.com/4780/109054677-0d1cc400-76d6-11eb-8622-4d753366f27d.PNG)  | ![IMG_0036](https://user-images.githubusercontent.com/4780/109054663-0aba6a00-76d6-11eb-8727-8d0d85d18f67.PNG)  |
|  ![IMG_0032](https://user-images.githubusercontent.com/4780/109054675-0c842d80-76d6-11eb-8bdf-beead1f6df1a.PNG) | ![IMG_0037](https://user-images.githubusercontent.com/4780/109054659-0a21d380-76d6-11eb-8876-1a83018beb93.PNG)  |
| ![IMG_0033](https://user-images.githubusercontent.com/4780/109054674-0beb9700-76d6-11eb-8924-6b1d1e3d2fed.PNG) | ![IMG_0038](https://user-images.githubusercontent.com/4780/109054654-09893d00-76d6-11eb-8487-e133640447a5.PNG) |
| ![IMG_0034](https://user-images.githubusercontent.com/4780/109054669-0b530080-76d6-11eb-9a31-c1d17410e745.PNG) | ![IMG_0039](https://user-images.githubusercontent.com/4780/109054651-09893d00-76d6-11eb-88d4-96779f2f33e7.PNG) |
| ![IMG_0035](https://user-images.githubusercontent.com/4780/109054667-0b530080-76d6-11eb-82bd-a25e236836a5.PNG) | ![IMG_0040](https://user-images.githubusercontent.com/4780/109054649-08581000-76d6-11eb-9121-755853a89d37.PNG) |

**To test**

- Enable the milestone notifications feature flag and build and run.
- View as many types of notification as you can and ensure they look okay. Check both 'badge' style notifications like those shown in the screenshots above, and other types like comments and likes to ensure they still look unchanged.
- Re-test with the feature flag disabled and check that the existing styles are still shown.

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

frosty committed:
Notification details: Add confetti background for milestones

frosty committed:
Notifications: Update fonts for badge style notifications to serif title1

frosty committed:
Merge branch 'feature/milestone-notifications-detail-button' into feature/milestone-notifications-styles

frosty committed:
Update NoteBlockTextTableViewCell margins for 'title' block vs others

frosty committed:
Badge notifications: Apply title style only to first text block

frosty committed:
Merge branch 'develop' into feature/milestone-notifications-styles

mbshakti commented:
Looks awesome. Thanks @frosty

Gio2018 commented on WordPress/Classes/Extensions/UIFont+Styles.swift:
```diff
@@ -0,0 +1,25 @@
+extension UIFont {
+    /// Returns a UIFont instance with the italic trait applied.
+    func italic() -> UIFont {
+        return withSymbolicTraits(.traitItalic)
+    }
+
+    /// Returns a UIFont instance with the semibold trait applied.
+    func semibold() -> UIFont {
```
nit: how about we use computed properties instead of methods?

frosty replied:
I kind of preferred methods as they're creating new font objects and not just accessing values of the existing font – and it felt a little SwiftUI-ish 🙂

Gio2018 commented on WordPress/Classes/Extensions/UIFont+Styles.swift:
```diff
@@ -0,0 +1,25 @@
+        return withWeight(.semibold)
+    }
+
+    func withSymbolicTraits(_ traits: UIFontDescriptor.SymbolicTraits) -> UIFont {
+        guard let descriptor = fontDescriptor.withSymbolicTraits(traits) else {
+            return self
+        }
+
+        return UIFont(descriptor: descriptor, size: 0)
+    }
+
+    func withWeight(_ weight: UIFont.Weight) -> UIFont {
```
Do we need to expose these two methods or can they be private?

frosty replied:
Yeah, these could be private!

Gio2018 reviewed (approved):
Just a couple of minor comments that can also be ignored, other than that it looks great!

:shipit: !

frosty committed:
UIFont+Styles: Make methods private that don't need to be exposed

frosty committed:
Fix a crash in note block comment cells.

* Note block comment cells subclass note block text cells, but don't
  have horizontal constraints, which would cause them to crash when the
  cell attempts to access the constraints.
